### PR TITLE
HM-10 BT driver

### PR DESCRIPTION
**Pull request by forGGe**
Implements #203

### CODE REVIEW

File: dev/bus/export/dev/bus.hpp [hunk 1]
```diff
@@ -335,6 +355,12 @@ void generic_bus<PBus>::unlock()
 template<class PBus>
 ecl::err generic_bus<PBus>::set_buffers(const uint8_t *tx, uint8_t *rx, size_t size)
 {
+    return set_buffers(tx, rx, size, size);
```
Comment (RostakaGmfun) [inline dev/bus/export/dev/bus.hpp]: Use overloaded version with 3 arguments

File: dev/hm10/export/dev/hm10.hpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+#include <algorithm>
+#include <chrono>
+
+namespace ecl
+{
+
+//! HM10 synchronous BT driver.
+//! \details This HM10 driver is simple and low-functional API to send and receive
+//! data from BT module. It lacks connection status detection and do not escape
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: does not

File: dev/hm10/export/dev/hm10.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+//! \note Driver _is_not_ a thread safe. Using driver routine from ISR context
+//! will lead to undefined behaviour.
+//! \tparam UART bus to work with.
+template<typename Uart>
+class hm10_sync
+{
+public:
+    //! Initializes BT module.
+    //! \return Status of operation.
+    //! \retval err::inval      Module respond with invalid data.
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: responded

File: dev/hm10/export/dev/hm10.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    static err init();
+
+    //! Sends data to the module.
+    //! \warning It is possible that no one has connected to the device yet.
+    //! \warning Data that is sent to unconnected module will be lost.
+    //! \details Call blocks until all data will be sent.
+    //! \param[in] buf Buffer to send.
+    //! \param[in] sz  Size of buffer.
+    //! \return Status of operation.
+    //! \retval err::ok         Module is ready to work.
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: Copy-paste?

File: dev/hm10/export/dev/hm10.hpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    static err send(const uint8_t *buf, size_t sz);
+
+    //! Receives data from the module with given timeout.
+    //! \param[out]     buf Buffer to store incoming data.
+    //! \param[in,out]  sz  Size of the buffer as input, bytes effectively written
+    //!                     as output.
+    //! \param[in]      ms  Time to wait for incoming data.
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: I propose to document that the function will block infinitely if the default value for `ms` argument is used.

File: dev/hm10/export/dev/hm10.hpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    //! \param[in] exp_resp Expected response.
+    //! \param[in] exp_sz   Size of expected response.
+    //! \param[in] ms       Time to wait for response.
+    //! \return Status of operation.
+    //! \retval err::timedout   Response is not received within given timeout.
+    //! \retval err::inval      Response is received but it is invalid.
+    //! \retval err::ok         Command succeed.
+    //! \todo Use uint8_t instead of char*.
+    static err send_cmd(const char *cmd, size_t cmd_sz, const char *exp_resp, size_t exp_sz,
+                    std::chrono::milliseconds ms = std::chrono::milliseconds(1000));
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: Let's move the magic timeout to the constant.

File: dev/hm10/export/dev/hm10.hpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    //! \retval err::timedout   Module did not respond at all.
+    //! \retval err::ok         Module is ready to work.
+    static err init();
+
+    //! Sends data to the module.
+    //! \warning It is possible that no one has connected to the device yet.
+    //! \warning Data that is sent to unconnected module will be lost.
+    //! \details Call blocks until all data will be sent.
+    //! \param[in] buf Buffer to send.
+    //! \param[in] sz  Size of buffer.
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: I propose to accept reference in order to update it with the actual number of bytes written.

File: dev/hm10/export/dev/hm10.hpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    char resp[32] = {0};
+
+    // TODO: provide better buffer management.
+    ecl_assert(sizeof(resp) > exp_sz);
+
+    size_t resp_sz = exp_sz;
+
+    auto rc = xfer(reinterpret_cast<const uint8_t*>(cmd),
+                   reinterpret_cast<uint8_t*>(resp),
+                   cmd_sz, resp_sz, ms);
```
Comment (RostakaGmfun) [inline dev/hm10/export/dev/hm10.hpp]: The `resp` array might be overflown if `resp_sz` > `32`

File: platform/stm32/export/aux/usart_bus.hpp [hunk 9]
```diff
@@ -591,11 +591,12 @@ void usart_bus<dev>::irq_handler()
             m_rx[m_rx_size - m_rx_left--] = static_cast<uint8_t>(data);
 
             if (listen_mode()) {
-                // Notify about state of the buffer.
-                event_handler()(channel::rx, event::tc, m_rx_size - m_rx_left);
-            }
+                // If buffer is full - start the next round
+                m_rx_left = m_rx_left ? m_rx_left : m_rx_size;
```
Comment (RostakaGmfun) [inline platform/stm32/export/aux/usart_bus.hpp]: Badly resolved merge conflict :(

File: dev/hm10/export/dev/hm10.hpp (hunk 7 above)
Comment (forGGe) [inline dev/hm10/export/dev/hm10.hpp]: Done

File: dev/hm10/export/dev/hm10.hpp (hunk 4 above)
Comment (forGGe) [inline dev/hm10/export/dev/hm10.hpp]: Done

File: dev/hm10/export/dev/hm10.hpp (hunk 5 above)
Comment (forGGe) [inline dev/hm10/export/dev/hm10.hpp]: Done

File: dev/hm10/export/dev/hm10.hpp (hunk 6 above)
Comment (forGGe) [inline dev/hm10/export/dev/hm10.hpp]: Done

**Comment by forGGe** [comment]
@RostakaGmfun please re-review

**Review by RostakaGmfun — APPROVED** [review]
Looks good!

**Pull request merged**